pkg/issuer/acme/http/http.go

```diff
@@ -57,27 +57,36 @@ var (
 type Solver struct {
 	*controller.Context
 
-	podLister     corev1listers.PodLister
-	serviceLister corev1listers.ServiceLister
-	ingressLister networkingv1beta1listers.IngressLister
+	podLister            corev1listers.PodLister
+	serviceLister        corev1listers.ServiceLister
+	ingressLister        ingress.InternalIngressLister
+	ingressCreateUpdater ingress.InternalIngressCreateUpdater
 
 	testReachability reachabilityTest
 	requiredPasses   int
 }
 
 type reachabilityTest func(ctx context.Context, url *url.URL, key string) error
 
-// NewSolver returns a new ACME HTTP01 solver for the given Issuer and client.
-// TODO: refactor this to have fewer args
-func NewSolver(ctx *controller.Context) *Solver {
-	return &Solver{
-		Context:          ctx,
-		podLister:        ctx.KubeSharedInformerFactory.Core().V1().Pods().Lister(),
-		serviceLister:    ctx.KubeSharedInformerFactory.Core().V1().Services().Lister(),
-		ingressLister:    ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Lister(),
-		testReachability: testReachability,
-		requiredPasses:   5,
+// NewSolver returns a new ACME HTTP01 solver for the given *controller.Context.
+func NewSolver(ctx *controller.Context) (*Solver, error) {
+	ingressLister, _, err := ingress.NewListerInformer(ctx)
+	if err != nil {
+		return nil, err
 	}
+	ingressCreateUpdater, err := ingress.NewCreateUpdater(ctx)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We can frequently end up doing 4 discovery API calls here. That's probably fine but. can we reduce the number to 2 by either passing in a result, or merging `NewCreateUpdater` `NewListerInformer` together?

Reply:
Perhaps we can even cache the result in a private global var in that package. We'd only need to do it once then, while keeping the same signatures.

Reply:
I've put a cache in the Internal Ingress package as suggested